### PR TITLE
Docs for cluster tiering to improve query concurrency

### PR DESCRIPTION
**Pull request by vtlim**
This PR adds a new doc describing query laning, Historical tiering, and Broker tiering.

This PR has:
- [x] been self-reviewed.

### CODE REVIEW

File: docs/operations/query-concurrency.md [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+  ~ under the License.
+  -->
+
+If you frequently run concurrent, mixed workloads on your Druid cluster, configure Druid to properly allocate cluster resources and optimize your overall query performance. With proper resource isolation, you can execute long-running, low priority queries that are resource intensive without interfering with short-running, high priority queries that require fewer resources. By separating cluster resources, you prevent queries from competing with each other for resources such as CPU, memory, and network access.
+
+There are two approaches to isolate your resources for improving query concurrency: query laning and cluster tiering. Use query laning to set a limit on the maximum number of long-running queries executed on each Broker. Use cluster tiering to define separate groups of Historicals and Brokers to which different queries can be directed based on their priority.
+
+## Query laning
+
+Query laning directs Druid to restrict resource usage for less urgent queries to ensure dedicated resources for higher priority queries. Query laning is ideal when you need to run many concurrent queries having heterogeneous workloads.
```
Comment (jihoonson) [inline docs/operations/query-concurrency.md:32]: The ideal solution to support heterogeneous workloads is the broker tiering today. Perhaps we should move this statement to the tiering section below?

File: docs/operations/query-concurrency.md [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+Query lanes are analogous to carpool and normal lanes on the freeway. With query laning, Druid restricts low priority queries to low lanes and allows high priority queries to run wherever possible, whether in a high or low lane. In this way, higher priority queries may bypass other queries in lower priority lanes.
+
+In Druid, query lanes reserve resources for Broker HTTP threads. Each Druid query requires one Broker thread. The number of threads on a Broker is defined by the `druid.server.http.numThreads` parameter. Broker threads may be occupied by tasks other than queries, such as health checks. You can use query laning to limit the number of HTTP threads designated for resource-intensive queries, leaving other threads available for short-running queries and other tasks.
+
+### General properties
+
+Set the following query laning properties in the `broker/runtime.properties` file.
+
+* `druid.query.scheduler.numThreads` – The total number of queries that can be served per Broker. We recommend setting this value to 1-2 less than `druid.server.http.numThreads`.
```
Comment (jihoonson) [inline docs/operations/query-concurrency.md:42]: Should this property appear after `laning.strategy` because it's optional?

File: docs/operations/query-concurrency.md [hunk 3]
```diff
@@ -0,0 +1,174 @@
+---
+id: query-concurrency
+title: Query concurrency
```
Comment (jihoonson) [inline docs/operations/query-concurrency.md:3]: I feel there could be a better title than this.. Would "Multi-workloads support" be better?

File: docs/operations/query-concurrency.md [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+}
+```
+
+See [Load rules](rule-configuration.md#load-rules) for more information on segment load rules. Visit [Tutorial: Configuring data retention](../tutorials/tutorial-retention.md) for an example of setting retention rules from the Druid web console.
+
+### Assign Historicals to tiers
+
+Assign the Historical to tiers by labeling the tier name and setting the priority value in the  `historical/runtime.properties` files.
+
+Example config for a Historical in the hot tier:
```
Comment (techdocsmith) [inline docs/operations/query-concurrency.md:110]:
```suggestion
Example Historical in the hot tier:
```

File: docs/operations/query-concurrency.md [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+```
+druid.server.tier=hot
+druid.server.priority=1
+```
+
+Example config for a Historical in the cold tier:
```
Comment (techdocsmith) [inline docs/operations/query-concurrency.md:117]:
```suggestion
Example Historical in the cold tier:
```

File: docs/operations/query-concurrency.md [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+```
+druid.server.tier=_default_tier
+druid.server.priority=0
+```
+
+See [Historical general configuration](../configuration/index.md#historical-general-configuration) for more details on these properties.
+
+## Broker tiering
+
+To set up Broker tiering, assign Brokers to tiers and configure query routing by the Router. You must set up Historical tiering before you can use Broker tiering.
```
Comment (techdocsmith) [inline docs/operations/query-concurrency.md:128]: Why would I do Broker tiering? What is the added benefit. Consider stating that it's dependent on Historical tiering first.

**Review by techdocsmith — CHANGES_REQUESTED** [review]
You're off to a great start here. I feel like the main work is to help the user understand when/why to choose one option over the other and what the benefits/pitfalls are for each.

File: docs/operations/query-concurrency.md (hunk 2 above)
Comment (vtlim) [reply docs/operations/query-concurrency.md:45]: reordered

File: docs/operations/query-concurrency.md [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+
+There are two approaches to isolate your resources for improving query concurrency: query laning and cluster tiering. Use query laning to set a limit on the maximum number of long-running queries executed on each Broker. Use cluster tiering to define separate groups of Historicals and Brokers to which different queries can be directed based on their priority.
+
+## Query laning
+
+Query laning directs Druid to restrict resource usage for less urgent queries to ensure dedicated resources for higher priority queries. Query laning is ideal when you need to run many concurrent queries having heterogeneous workloads.
+
+Query lanes are analogous to carpool and normal lanes on the freeway. With query laning, Druid restricts low priority queries to low lanes and allows high priority queries to run wherever possible, whether in a high or low lane. In this way, higher priority queries may bypass other queries in lower priority lanes.
```
[the remark this answers is not in the document]
Comment (vtlim) [reply docs/operations/query-concurrency.md:36]: @jihoonson Brokers decide which queries go into which lanes by query prioritization, right? 
I mentioned this in L49 but maybe I need to emphasize this better.

File: docs/operations/query-concurrency.md [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+# Laning strategy
+druid.query.scheduler.laning.strategy=hilo
+druid.query.scheduler.laning.maxLowPercent=20
+```
+
+See [Query prioritization and laning](../configuration/index.md#query-prioritization-and-laning) for details on query laning and the available query laning strategies.
+
+## Historical tiering
+
+Configure segment loading and tiers for Historicals as described in this section. In the examples below, we set up two tiers—hot and cold—for the Historicals and for the Brokers. We’ll instruct the Brokers to serve hot queries before cold queries. Short-running queries will be routed to the hot tiers, and long-running queries will be routed to the cold tiers.
```
[the remark this answers is not in the document]
Comment (vtlim) [reply docs/operations/query-concurrency.md:80]: good points. along with the unclear terminology on high and low lanes, I'll introduce these terms explicitly beforehand

File: docs/operations/query-concurrency.md [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+  ~
+  ~ Unless required by applicable law or agreed to in writing,
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied.  See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+
+If you frequently run concurrent, mixed workloads on your Druid cluster, configure Druid to properly allocate cluster resources and optimize your overall query performance. With proper resource isolation, you can execute long-running, low priority queries that are resource intensive without interfering with short-running, high priority queries that require fewer resources. By separating cluster resources, you prevent queries from competing with each other for resources such as CPU, memory, and network access.
```
[the remark this answers is not in the document]
Comment (vtlim) [reply docs/operations/query-concurrency.md]: you're correct -- you typically do not want the long resource-intensive queries to throttle the performance of short interactive queries that require fewer resources. I'll add that and make the other changes to the text.

File: docs/operations/query-concurrency.md (hunk 1 above)
Comment (vtlim) [reply docs/operations/query-concurrency.md]: Synced up with Jihoon and we'll leave query laning first. I added a new paragraph to describe the distinction on when to use which approach.

File: docs/operations/query-concurrency.md (hunk 6 above)
Comment (vtlim) [reply docs/operations/query-concurrency.md]: added content to top of "Service tiering" section

File: docs/operations/mixed-workloads.md [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+# This value should be less than druid.server.http.numThreads
+druid.query.scheduler.numThreads=40
+```
+
+
+## Service tiering
+
+In service tiering, you define separate groups of Historicals and Brokers to manage queries based on the segments and resource requirements of the query.
+You can limit the resources that are set aside for certain types of queries.
+Many heavy queries involving complex subqueries or large result sets can crash a Broker, or worse, take down a cluster.
```
Comment (jihoonson) [inline docs/operations/mixed-workloads.md:100]: This sounds a bit scarier than necessary as it technically can happen but is not common. Can we mention the problem of resource contention between long-running queries and short-running queries instead?

File: docs/operations/mixed-workloads.md [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+Minimize the impact of these heavy queries by limiting them to a separate Broker tier.
+When all Brokers set aside for heavy queries are occupied, users must wait to submit additional heavy queries until the designated resources become available.
```
Comment (jihoonson) [inline docs/operations/mixed-workloads.md:102]: This looks a bit unclear to me. I think it's because there is nothing in Druid disallowing users to submit additional queries. Perhaps 

```suggestion
When all Brokers set aside for heavy queries are occupied, additional queries submitted subsequently can wait until the designated resources become available, which can fail those queries after query timeout expires.
```

Comment (vtlim) [reply docs/operations/mixed-workloads.md:102]:
```suggestion
When all Brokers set aside for heavy queries are occupied, subsequent queries must wait until the designated resources become available.
A prolonged wait results in the later queries failing with a timeout error.
```

File: docs/operations/mixed-workloads.md (hunk 10 above)
Comment (vtlim) [reply docs/operations/mixed-workloads.md:100]:
```suggestion
Many heavy queries involving complex subqueries or large result sets can hog resources away from high priority, interactive queries.
```

File: docs/operations/mixed-workloads.md [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+You can profile Druid queries using normal performance profiling techniques such as Druid query metrics analysis, thread dumps of JVM, or flame graphs to identify what resources are affected by mixed workloads.
+The largest bottleneck will likely be in the Broker HTTP threads.
+Mitigate resource contention of the Broker HTTP threads with query laning.
+However, mixed workloads also affect other resources, including processing threads and merge buffers.
+Reduce the burden on these resources by applying service tiering in addition to query laning.
+
+
+## Query laning
+
+Query laning directs Druid to restrict resource usage for less urgent queries to ensure dedicated resources for high priority queries. Query laning is ideal when you need to run many concurrent queries having heterogeneous workloads.
```
[the remark this answers is not in the document]
Comment (vtlim) [reply docs/operations/mixed-workloads.md:48]:
```suggestion
When you need to run many concurrent queries having heterogeneous workloads, start with query laning to optimize your query performance.
Query laning restricts resource usage for less urgent queries to ensure dedicated resources for high priority queries.
```

**Review by jihoonson — APPROVED** [review]
LGTM. Thank you @vtlim, this doc will be very useful for our users!

File: docs/operations/mixed-workloads.md [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied.  See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+
+If you frequently run concurrent, heterogeneous workloads on your Apache Druid cluster, configure Druid to properly allocate cluster resources to optimize your overall query performance.
+
+Each Druid query consumes a certain amount of cluster resources, such as processing threads, memory buffers for intermediate query results, and HTTP threads for communicating between Brokers and data servers.
```
Comment (techdocsmith) [inline docs/operations/mixed-workloads.md:28]: Great job describing  the "heavy" vs "light" here.
nit: Do we need the carriage return at the end of line 28?

File: docs/operations/mixed-workloads.md [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+
+Druid provides the following strategies to isolate resources and improve query concurrency:
+- **Query laning** where you set a limit on the maximum number of long-running queries executed on each Broker. 
+- **Service tiering** which defines separate groups of Historicals and Brokers to receive different query assignments based on query priority.
+
+You can profile Druid queries using normal performance profiling techniques such as Druid query metrics analysis, thread dumps of JVM, or flame graphs to identify what resources are affected by mixed workloads.
+The largest bottleneck will likely be in the Broker HTTP threads.
+Mitigate resource contention of the Broker HTTP threads with query laning.
+However, mixed workloads also affect other resources, including processing threads and merge buffers.
+Reduce the burden on these resources by applying service tiering in addition to query laning.
```
Comment (techdocsmith) [inline docs/operations/mixed-workloads.md:43]:
```suggestion
To reduce the burden on these resources, apply both service tiering and to query laning.
```
nit
Also not sure about the carriage returns in this 'graf

File: docs/operations/mixed-workloads.md [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+
+
+## Query laning
+
+When you need to run many concurrent queries having heterogeneous workloads, start with query laning to optimize your query performance.
+Query laning restricts resource usage for less urgent queries to ensure dedicated resources for high priority queries.
+
+Query lanes are analogous to carpool and normal lanes on the freeway. With query laning, Druid sets apart VIP lanes from other general lanes.
```
Comment (techdocsmith) [inline docs/operations/mixed-workloads.md:51]: Wonder if there is another term than VIP (non-native speakers may trip over that one)

File: docs/operations/mixed-workloads.md [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+Minimize the impact of these heavy queries by limiting them to a separate Broker tier.
+When all Brokers set aside for heavy queries are occupied, subsequent heavy queries must wait until the designated resources become available.
+A prolonged wait results in the later queries failing with a timeout error.
+
+Note that you can separate Historical processes into tiers without having separate Broker tiers.
+Historical-only tiering is not sufficient to meet the demands of mixed workloads on a Druid cluster.
+However, it is useful when you query certain segments more frequently than others, such as often analyzing the most recent data.
+Historical tiering assigns data from specific time intervals to specific tiers in order to support higher concurrency on hot data. 
+
+The examples below demonstrate two tiers—hot and cold—for both the Historicals and Brokers. The Brokers will serve short-running, light queries before long-running, heavy queries. Light queries will be routed to the hot tiers, and heavy queries will be routed to the cold tiers.
```
Comment (techdocsmith) [inline docs/operations/mixed-workloads.md:111]: I notice that we switch to hot and cold here. Are these related to "heavy" and "light" introduced above?

Comment (techdocsmith) [reply docs/operations/mixed-workloads.md:111]: Not for this PR, but I would like to see an example of how  two queries, one heavy/cold or light/hot and how they would be routed using this strategy. I think that might be a good result of the research of using the flame graphs, etc.

**Review by techdocsmith — CHANGES_REQUESTED** [review]
Few small changes for this PR, but some ideas for clarity in a future revision.

File: docs/operations/mixed-workloads.md (hunk 13 above)
Comment (vtlim) [reply docs/operations/mixed-workloads.md:28]: the single carriage return leaves the content in the same paragraph so I like to have it to make it easier to track future changes

File: docs/operations/mixed-workloads.md (hunk 15 above)
Comment (vtlim) [reply docs/operations/mixed-workloads.md:51]:
```suggestion
Query lanes are analogous to carpool and normal lanes on the freeway. With query laning, Druid sets apart particular lanes from other general lanes.
```

File: docs/operations/mixed-workloads.md (hunk 16 above)
Comment (vtlim) [reply docs/operations/mixed-workloads.md:111]: In this doc I'm using "hot/cold" for data and tiers and "heavy/light" for describing the queries. The crucial bit that connects the two concepts is in L111: "Light queries will be routed to the hot tiers, and heavy queries will be routed to the cold tiers." Is that clear enough?

File: docs/operations/mixed-workloads.md [hunk 17]
```diff
@@ -40,15 +40,15 @@ You can profile Druid queries using normal performance profiling techniques such
 The largest bottleneck will likely be in the Broker HTTP threads.
 Mitigate resource contention of the Broker HTTP threads with query laning.
 However, mixed workloads also affect other resources, including processing threads and merge buffers.
-Reduce the burden on these resources by applying service tiering in addition to query laning.
+To reduce the burden on these resources, apply both service tiering and to query laning.
```
Comment (vtlim) [inline docs/operations/mixed-workloads.md:43]:
```suggestion
To reduce the burden on these resources, apply both service tiering and query laning.
```

File: docs/operations/mixed-workloads.md (hunk 16 above)
Comment (techdocsmith) [reply docs/operations/mixed-workloads.md:111]: Sounds good.

**Review by techdocsmith — APPROVED** [review]
LGTM

**Review by asdf2014 — APPROVED** [review]
:+1: LGTM, thanks a lot for your contribution
